### PR TITLE
fix: replicate memtable mutations via custom rmgr

### LINKED ISSUE #345: bug: stale results & errors on streaming-standby for persistent connections
## Summary

On a streaming physical standby, pg_textsearch backends that **stay alive across primary inserts** (connection pools, app servers with persistent connections, persistent psql sessions) silently return stale results after the primary writes new data. Disconnecting and reconnecting masks the problem.

This is a distinct bug class from #342 (which was about WAL silence on the primary side, fixed by PR #343). After #343, the on-disk index correctly replicates to the standby; this issue is about per-backend in-memory state on the standby not noticing the on-disk changes.

## Motivating workload

The canonical scaling pattern for read-heavy workloads is to route writes to the primary and reads to one or more streaming replicas behind a connection pool. Under that pattern, every long-lived pooled connection on the replica is exposed to this bug.

## Why it happens

- On the primary, every INSERT goes through pg_textsearch's `aminsert` callback, which appends to the in-memory memtable (DSA-backed shared structure) and bumps the docid pages. Inserts are visible immediately because the same backend that wrote them is also reading the live memtable.
- On the standby, INSERTs aren't executed — they arrive as WAL records that the startup process applies to the on-disk index (docid pages, segment pages, metapage). **WAL replay never calls into `aminsert`**, so the in-memory memtable on the standby is not updated by replication.
- The infrastructure to rebuild the memtable from on-disk state (`tp_rebuild_posting_lists_from_docids`, which re-tokenizes from the heap) does exist and runs lazily on a backend's first scan when it has no shared registry entry. So a *fresh* standby connection sees current data — which is why one-shot `psql -c` queries appear to work and why the bug stayed hidden.
- A *long-lived* standby backend rebuilds once on first scan and then never refreshes, so subsequent primary inserts are invisible to it. The same per-backend cache also pins a snapshot of the segment list, so post-merge segment changes are not picked up either.

## Failure modes observed in the test suite

1. **Memtable staleness** — standby backend misses primary inserts that arrived as WAL after the backend's first scan. Queries return 0 results (or undercount) for newly-added documents.
2. **Post-merge invisibility** — after the primary spills/merges, the long-lived standby backend's cached segment list still points at the pre-merge state. Queries return 0 results for documents that have moved into a segment the backend doesn't know about.

(Before PR #343 there was also a third surface, `ERROR: invalid page index` / `could not read blocks`, but that was a symptom of #342 — page-index pages literally not reaching the standby's on-disk index. Fixed.)

## Reproducers in the suite

- `test/scripts/replication.sh` test 5 — single long-lived standby session, before/after a primary insert.
- `test/scripts/replication_correctness.sh` A2–A5, E1–E5 — long-lived standby backends across UPDATE / DELETE+VACUUM / partitioned / expression / partial / array / multi-index variations.
- `test/scripts/replication_concurrency.sh` B2 — eight concurrent long-lived standby readers under primary write load.
- `test/scripts/replication_cascading.sh` C2 — bug propagates two replication hops (primary → standby1 → standby2).
- `test/scripts/replication_compat.sh` F2 — same surface inside a logical+physical coexistence setup.

All currently fail with `BUG (expected): ...` messages keyed to this issue.

## Fix options

**Option A — custom rmgr (preferred).** Define pg_textsearch-owned WAL record types (`INSERT_TERMS`, `SPILL`, `MERGE`, etc.) and `redo` callbacks that mutate the standby's shared-memory memtable directly during recovery. Architecturally honest; the standby memtable becomes a real replica of the primary's. Costs: larger primary WAL footprint (term/freq pairs per doc), and the redo path has to be compatible with the startup process's constraints (runs before normal backends are up).

**Option B — lazy detect-and-refresh on the standby read path.** Track a "rebuilt-through" cursor `(last_docid_blkno, slot_in_page)`. On each scan, when `RecoveryInProgress()`, compare the docid-chain tail to the cursor and re-tokenize just the new TIDs from the heap. On primary spill, the replicated metapage update is the signal to clear the in-memory memtable; subsequent reads see the new on-disk segment plus an empty memtable. Reuses existing rebuild code, no new WAL volume, but pays per-scan refresh cost on the replica — which is exactly the wrong tradeoff for read-from-replica scaling workloads.

A is the right answer for this issue's motivating workload; B is a fallback if WAL-volume cost on the primary turns out to be unacceptable.

### PR DESCRIPTION
**Pull request by tjgreen42**
Closes #345, #349, #350.

## What was broken

`pg_textsearch` keeps the working set of recent docs in a DSA shared-memory memtable (one per index, shared across backends). Pre-PR, the memtable on a streaming standby was populated only by the standby's own first-access rebuild. After that, primary writes did not propagate: they updated docid pages on disk via GenericXLog but never touched the standby's in-memory memtable. Three bugs followed.

- **#345**: a long-lived standby backend's BM25 query results were frozen at the time of its first index access. Subsequent primary inserts were invisible until the backend reconnected.
- **#349**: when the primary compacted L0 segments into L1, the standby received the metapage unlink/link via GenericXLog (buffer locks only), but a backend mid-walk on the L0 chain followed `next_segment` pointers into pages that the FSM had since recycled. Symptoms ranged from wrong counts to `ERROR: invalid page index at block N / magic=0x00000000`.
- **#350**: `total_docs` and `total_len` (the BM25 corpus stats that drive IDF and BMW) were never WAL-logged. After PITR to a between-spills LSN, the recovered cluster's metapage held the last spill's snapshot, the in-memory atomics started at zero, and the standard query path returned zero results until a `REINDEX`.

## Summary

A custom WAL resource manager (rmgr ID 149, [registered](https://wiki.postgresql.org/wiki/CustomWALResourceManagers)) emits three record types so streaming-replication standbys and primary crash recovery keep their in-shared-memory memtable in sync with the segment chain on disk.

| Record | Standby redo |
|---|---|
| `INSERT_TERMS` | Apply the v2 bm25vector to the memtable (idempotent by CTID). |
| `SPILL` | Apply the L0 chain-link, optional new-segment splice, and memtable clear, atomically under per-index `LW_EXCLUSIVE`. |
| `MERGE_LINKAGE` | Apply the metap unlink/link, optional new-segment splice, and corpus-stat shrinkage, atomically under `LW_EXCLUSIVE`. |

`SPILL` is emitted from all five spill recipes (auto-spill, `bm25_spill_index`, VACUUM bulkdelete, PRE_COMMIT bulk-load, global-memory eviction). Each spill entry point also has a `RecoveryInProgress()` guard so a standby never autonomously spills. Emission is gated on `RelationNeedsWAL(rel)` so UNLOGGED and TEMP indexes do not ship records that cannot apply on a standby.

The full invariants and scenario-by-scenario verification of the cold-start bootstrap path are spec'd at [`docs/replication/CORRECTNESS.md`](docs/replication/CORRECTNESS.md). Read that before changing `tp_rebuild_index_from_disk`.

## Notes on the tricky parts

- **Corpus stats.** The bootstrap writes `total_docs = metap.total + memtable_count` via `pg_atomic_write` under the per-index exclusive lock, not `fetch_add`. The two halves are disjoint by construction (segments versus the memtable), so the write is unambiguous; `fetch_add` would double-count any cohort that redo applied during the WAL-replay drain and that a subsequent SPILL also folded into `metap.total_docs`. The PITR "BM25 returns zero" symptom in #350 was the pre-fix unconditional-overwrite case.

- **Concurrent bootstrap.** Two backends arriving at the same empty registry slot resolve to the same shared state via `tp_registry_register_if_absent` (atomic register-or-attach), not to a unregister/re-register race.

- **Defensive WAL decoding.** Header-length checks before any field deref; bounds check on `MERGE_LINKAGE.level` before mutating `metap.level_heads/level_counts`; designated initializers on all emission headers so compiler padding is zeroed.

- **`bm25_force_merge` locking.** `tp_xlog_merge_linkage` requires the caller to hold the per-index `LW_EXCLUSIVE`; this serializes the standby's redo against in-flight scans (#349). All other call paths take that lock first, but `bm25_force_merge` was a pre-existing exception: it held only `RowExclusiveLock` on the relation and called the merge helper directly. The SQL function now wraps the force-merge body in `tp_acquire_index_lock(LW_EXCLUSIVE)` / `tp_release_index_lock`.

## Testing

| Suite | New coverage |
|---|---|
| `replication_spill_paths.sh` | 4 tests: `tp_xlog_spill` from each spill recipe, cold-start corpus stats (#350), `MERGE_LINKAGE` round-trip (#349) |
| `wal_audit.sh` | 3 tests: `RelationNeedsWAL` guard, `tp_rmgr_identify` names, `INSERT_TERMS` records present in WAL window |
| `recovery.sh` | `test_empty_index_rebuild`: empty index + restart fires the bootstrap empty-index branch |
| `unlogged_index.sql` | UNLOGGED merge through `merge.c` GenericXLog fallback; delete + VACUUM + force-merge to drive shrinkage on that path |

Existing replication suite (`replication.sh`, `replication_correctness.sh`, `replication_concurrency.sh`, `replication_failover.sh`, `replication_cascading.sh`, `replication_compat.sh`, `replication_parallel_build.sh`, `replication_issue_342.sh`) all pass.

### CODE REVIEW

File: src/access/build.c [hunk 1]
```diff
@@ -1573,6 +1597,19 @@ tp_insert(
...
+			 * primary crash recovery) can apply the same term
+			 * list to its memtable. Critical section is required
+			 * by XLogInsert; the WAL must be issued while we
+			 * still hold the per-index lock so the order of
+			 * memtable mutations is consistent across primary
+			 * and standby.
+			 */
+			START_CRIT_SECTION();
+			tp_xlog_insert_terms(RelationGetRelid(index), ht_ctid, tpvec);
+			END_CRIT_SECTION();
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/access/build.c]: Fixed in 1aa830c2 — all six emission sites (`tp_do_spill`, `tp_spill_memtable`, `tp_insert` in `build.c`; `tp_spill_memtable_if_needed` in `vacuum.c`; `tp_evict_largest_memtable` in `registry.c`; `tp_bulk_load_spill_check` in `state.c`) now wrap the `tp_xlog_*` call in `if (RelationNeedsWAL(rel))`.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -553,6 +553,9 @@ jobs:
         (cd test/scripts && ./shutdown_spill.sh) || true
         (cd test/scripts && ./cic.sh) || true
         (cd test/scripts && ./memory_accounting.sh) || true
+        # replication.sh exercises the WAL rmgr redo paths
+        # (INSERT_TERMS, SPILL) on the standby's startup process.
+        (cd test/scripts && ./replication.sh) || true
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply .github/workflows/ci.yml:580]: Fixed in 1aa830c2 — `replication.sh` now runs in the regular `test` job with strict if/else failure handling. The tolerant invocation in the coverage job stays for gcda flushing.

File: src/replication/rmgr.c [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+	int32			   *frequencies = NULL;
+	int					term_count;
+	int					doc_length = 0;
+	int					i;
+
+	if (XLogRecGetDataLen(record) < sizeof(*hdr) + hdr->vector_size)
+		elog(PANIC,
+			 "INSERT_TERMS record truncated: declared %u, got %u",
+			 hdr->vector_size,
+			 (uint32)(XLogRecGetDataLen(record) - sizeof(*hdr)));
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/replication/rmgr.c:120]: Fixed in 1aa830c2 — split into a `< sizeof(*hdr)` check that runs before `hdr->vector_size` is dereferenced, with honest diagnostics in both failure modes.

File: src/replication/rmgr.c [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+
+	hdr.index_oid	= index_oid;
+	hdr.ctid		= *ctid;
+	hdr.vector_size = (uint32)vec_size;
+
+	XLogBeginInsert();
+	XLogRegisterData((char *)&hdr, sizeof(hdr));
+	XLogRegisterData((char *)vec, vec_size);
+	return XLogInsert(TP_RMGR_ID, XLOG_TP_INSERT_TERMS);
+}
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/replication/rmgr.c:582]: Fixed in 1aa830c2 — both `tp_xlog_insert_terms` and `tp_xlog_spill` now use designated initializers, which zero the padding under C99 §6.7.8/21.

File: src/index/state.c [hunk 5]
```diff
@@ -1057,17 +1098,18 @@ tp_rebuild_index_from_disk(Oid index_oid)
...
+			index_oid,
+			index_rel->rd_index->indrelid,
+			/* start_locked */ true);
 
 	if (local_state != NULL)
 	{
-		/* Rebuild posting lists from docid pages (if any) */
 		tp_rebuild_posting_lists_from_docids(index_rel, local_state, metap);
 
 		/*
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/index/state.c]: Fixed in d93118c3 — the rebuild path now drains WAL replay (waits for `GetXLogReplayRecPtr` to catch up to the wal-receiver's flush LSN) before walking docid pages, and the walk runs under `LW_SHARED`. `tp_add_document_terms` is now idempotent by CTID via a `dshash_find_or_insert` gate on the doc-length table, so concurrent redo + walk can no longer double-add or miss. `start_locked` is gone.

File: src/access/build.c [hunk 6]
```diff
@@ -139,6 +149,13 @@ tp_do_spill(TpLocalIndexState *index_state, Relation index_rel)
 	tp_link_l0_chain_head(index_rel, root);
 	tp_sync_metapage_stats(index_rel, index_state);
 
+	if (RelationNeedsWAL(index_rel))
+	{
+		START_CRIT_SECTION();
+		tp_xlog_spill(RelationGetRelid(index_rel));
+		END_CRIT_SECTION();
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/access/build.c]: Fixed in f550bbd6 — the L0 chain-link metapage update is now folded into the SPILL custom WAL record instead of being a separate GenericXLog record, so on the standby the chain-link mutation, the optional new-segment `next_segment` splice, and the memtable clear all replay atomically under per-index `LW_EXCLUSIVE` in `tp_redo_apply_spill`. No window where a backend holding `LW_SHARED` can observe the new segment queryable while the memtable still has the same docs.

File: src/index/state.c [hunk 7]
```diff
@@ -1060,26 +1077,78 @@ tp_rebuild_index_from_disk(Oid index_oid)
...
-		 * Load corpus statistics from metapage. This is needed for indexes
-		 * built with parallel workers (which write directly to segments
-		 * without docid pages) or if docid recovery didn't fully restore
-		 * the stats. The metapage is the authoritative source for total_docs
-		 * and total_len.
+		 * Add the metapage corpus stats to whatever the docid walk
+		 * recovered. Each source covers a disjoint slice of the index:
+		 *
+		 *   - metap->total_docs / total_len reflect docs that have
+		 *     been spilled into segments (tp_sync_metapage_stats runs
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/index/state.c]: Fixed in 9fcaf721 — `tp_rebuild_index_from_disk` now re-reads the metapage from disk after the WAL drain (and after acquiring `LW_EXCLUSIVE`) and uses those fresh values for both the docid walk and the corpus-stat `fetch_add`. The early metap read is kept only for the magic / heap-empty validation gates above, which are stable across replay.

[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/index/state.c]: Fixed in 9fcaf721 — the rebuild walk now holds `LW_EXCLUSIVE` across the docid walk + `tp_ensure_string_table_initialized` + `fetch_add`, satisfying the contract. The stale comment at tp_rebuild_posting_lists_from_docids is corrected to point at the new lock invariant. Deferred-downgrade variant rejected because the rebuild only fires on the first backend to open the index after server start — bounded duration, so the simpler "hold EXCLUSIVE throughout" pattern is sufficient.

File: src/replication/rmgr.c [hunk 8]
```diff
@@ -0,0 +1,765 @@
...
+	{
+		segbuf = ReadBuffer(index, new_segment_root);
+		LockBuffer(segbuf, BUFFER_LOCK_EXCLUSIVE);
+		segpage = BufferGetPage(segbuf);
+		/* Ensure pd_lower covers the contents (matches the
+		 * GenericXLog path tp_link_l0_chain_head used to take). */
+		((PageHeader)segpage)->pd_lower = BLCKSZ;
+		seg_header				 = (TpSegmentHeader *)PageGetContents(segpage);
+		seg_header->next_segment = prev_chain_head;
+	}
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/replication/rmgr.c:644]: Fixed in 9fcaf721 — `tp_xlog_spill` and `tp_xlog_merge_linkage` now manage their own critical section internally: ReadBuffer + LockBuffer happen first, then START_CRIT_SECTION, then the page mutations + MarkBufferDirty + XLogInsert + PageSetLSN, then END_CRIT_SECTION, then UnlockReleaseBuffer. The wrapping `START_CRIT_SECTION/END_CRIT_SECTION` at all six call sites (build.c x2, vacuum.c, registry.c, state.c, merge.c) is removed. Doc comment in replication.h updated.

File: src/replication/rmgr.c [hunk 9]
```diff
@@ -0,0 +1,765 @@
...
+
+	hdr.index_oid		 = RelationGetRelid(index);
+	hdr.level			 = level;
+	hdr.segment_count	 = segment_count;
+	hdr.total_at_level	 = total_at_level;
+	hdr.remainder_head	 = remainder_head;
+	hdr.new_segment		 = new_segment;
+	hdr.prev_target_head = prev_target_head;
+	hdr.docs_shrinkage	 = docs_shrinkage;
+	hdr.tokens_shrinkage = tokens_shrinkage;
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/replication/rmgr.c]: Fixed in 9fcaf721 — `tp_xlog_merge_linkage` now uses a designated initializer for its `xl_tp_merge_linkage hdr`, matching the `tp_xlog_insert_terms` / `tp_xlog_spill` pattern. C99 §6.7.8/21 zeros the padding bytes.

File: src/index/state.c [hunk 10]
```diff
@@ -1060,26 +1077,118 @@ tp_rebuild_index_from_disk(Oid index_oid)
...
+		pg_atomic_fetch_add_u32(
+				&local_state->shared->total_docs, fresh_metap->total_docs);
+		pg_atomic_fetch_add_u64(
+				&local_state->shared->total_len, fresh_metap->total_len);
+
+		pfree(fresh_metap);
+		tp_release_index_lock(local_state);
 	}
 
 	/* Clean up */
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/index/state.c]: Fixed in 3171759e — switched from `pg_atomic_fetch_add` to `pg_atomic_write` under LW_EXCLUSIVE. New `tp_doclength_summary` walks the doc-length dshash after the docid scan to compute the memtable cohort (count + total_len), then we set `shared->total = fresh_metap->total + memtable_cohort`. The memtable cohort is single-source-of-truth via the idempotency gate, so absolute write under EXCLUSIVE is safe.

File: src/index/state.c [hunk 11]
```diff
@@ -1060,26 +1077,118 @@ tp_rebuild_index_from_disk(Oid index_oid)
 				index_oid, index_rel->rd_index->indrelid);
 	}
 
-	/* Create fresh state first */
-	/* Creating fresh state for restart recovery */
 	local_state = tp_create_shared_index_state(
 			index_oid, index_rel->rd_index->indrelid);
 
 	if (local_state != NULL)
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/index/state.c]: Fixed in 3171759e — `tp_rebuild_index_from_disk` is restructured: magic check stays pre-register (stable), then register, drain, acquire LW_EXCLUSIVE, re-read metap, then run the heap-empty and empty-index decisions off the fresh values. The early-return-with-empty-state path now happens with a registered shared state under the lock, so any subsequent INSERT_TERMS / docid-page WAL applies against it correctly instead of being silently lost.

File: CLAUDE.md [hunk 12]
```diff
@@ -31,6 +31,23 @@ consider a dedicated `pg_textsearch` schema for cleaner namespace management.
...
+  `tp_clear_memtable`. The startup process can't open relations
+  (no transaction state), so redo bypasses `tp_get_local_index_state`
+  and constructs a minimal `TpLocalIndexState` on the stack from
+  the registry's shared state plus the global DSA. WAL emission
+  is gated on `RelationNeedsWAL(rel)` so UNLOGGED / TEMP indexes
+  don't pay for records that can't apply on the standby. Docid
+  pages remain the durable on-disk record of what's in the
+  memtable; the rebuild path in `tp_rebuild_index_from_disk`
+  registers its shared state with the per-index LWLock already
+  held EXCLUSIVE so a concurrent standby redo can't race the
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply CLAUDE.md]: Already addressed in 6d4a0ebd (committed before this review ran but bot may have snapshotted earlier). The current CLAUDE.md text describes the drain → acquire → re-read → walk flow plus the idempotency gate; the `start_locked` paragraph is gone.

**Comment by tjgreen42** [comment]
## Code Review Findings

### 🔴 High Severity: Out-of-bounds array access in MERGE_LINKAGE redo

**File:** `src/replication/rmgr.c` lines 413-417

**Problem:** The `tp_redo_apply_merge_linkage` handler doesn't validate `hdr->level` before using it as an array index. If a WAL record contains `hdr->level >= TP_MAX_LEVELS - 1` (where `TP_MAX_LEVELS = 8`), the code writes past the end of `level_heads` and `level_counts` arrays:

```c
metap->level_heads[hdr->level] = hdr->remainder_head;
metap->level_counts[hdr->level] = ...;
metap->level_heads[hdr->level + 1] = hdr->new_segment;  // ← +1 can overflow
metap->level_counts[hdr->level + 1]++;
```

**Why it matters:** While the primary code validates the level before emitting the record (merge.c:1430-1436), WAL redo must independently validate all fields. WAL records could come from:
- Corrupted archives
- Older/incompatible code versions
- Base backups with version drift

**Suggested fix:**
```c
if (hdr->level >= TP_MAX_LEVELS - 1)
    elog(PANIC, "MERGE_LINKAGE level %u exceeds TP_MAX_LEVELS-1", hdr->level);
```

---

### ✅ No Other Significant Issues Found

The rest of the implementation looks sound:
- Critical section handling follows Postgres conventions
- `RecoveryInProgress()` guards prevent standby autonomous spills
- `RelationNeedsWAL()` gates prevent UNLOGGED/TEMP WAL emission
- Corpus stats use `fetch_add` rather than overwrite (fixes #350)
- Cold-start rebuild properly drains WAL before walking docid pages
- Test coverage appears comprehensive (7 new targeted tests)

File: test/scripts/wal_audit.sh [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    fi
+
+    # Index must be functional post-recovery — first-access
+    # rebuild walks docid pages and populates the memtable. Verify
+    # via bm25_summarize_index rather than a BMW query (the BMW
+    # path checks shared->total_docs, which the rebuild path
+    # overwrites with a stale metap->total_docs of 0 when no spill
+    # has synced the metapage — orthogonal issue, see CLAUDE.md
+    # note).
+    local summary memtable_docs
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply test/scripts/wal_audit.sh:231]: Fixed in b5586403 — comment trimmed to the still-true rationale ("direct memtable-size assertion independent of BMW corner cases"). The pre-#350-fix overwrite behavior the old comment described is gone.

File: src/replication/replication.h [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+/*
+ * Emission helpers — called from the primary's insert/spill/merge
+ * paths. Return the LSN of the emitted record. The caller MUST
+ * already hold the per-index LW_EXCLUSIVE; the helpers manage
+ * their own buffer locks and critical section internally
+ * (acquiring buffers outside the CS so ReadBuffer ERROR doesn't
+ * escalate to PANIC).
+ */
+extern XLogRecPtr
+tp_xlog_insert_terms(Oid index_oid, ItemPointer ctid, const TpVector *vec);
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/replication/replication.h:154]: Fixed in b5586403 — split the shared doc-block into per-helper doc-blocks. tp_xlog_insert_terms doesn't require LW_EXCLUSIVE (caller holds either mode; helper body doesn't read/mutate shared state). tp_xlog_spill and tp_xlog_merge_linkage do require it (their redo also takes LW_EXCLUSIVE so backend scans serialize).

File: src/index/state.c [hunk 15]
```diff
@@ -970,16 +1002,33 @@ tp_cleanup_index_shared_memory(Oid index_oid)
...
+ * same memtable we are still rebuilding from docid pages — and
+ * because docid-page WAL is ordered after INSERT_TERMS WAL for any
+ * given insert, redo and the rebuild would both add the same doc
+ * (duplicate posting entries, inflated total_docs). To prevent
+ * that, we ask tp_create_shared_index_state to register the entry
+ * with the per-index LWLock already held EXCLUSIVE; the standby's
+ * single-threaded WAL replay simply blocks on that lock until the
+ * docid walk finishes and we release it, then proceeds with the
+ * normal redo path against the now-populated memtable.
  */
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/index/state.c:1030]: Fixed in b5586403 — docstring rewritten to summarize the current 8-step flow (validate magic → register-or-attach → drain → acquire LW_EXCLUSIVE → re-read metap → branch → walk + atomic_write → release) and link to docs/replication/CORRECTNESS.md for the full invariants and scenario walkthrough. The start_locked text is gone.

File: src/access/build.c [hunk 16]
```diff
@@ -494,10 +521,17 @@ tp_spill_memtable(PG_FUNCTION_ARGS)
...
+		{
+			tp_xlog_spill(index_rel, segment_root);
+		}
+		else
+		{
+			tp_link_l0_chain_head(index_rel, segment_root);
+		}
+
 		tp_maybe_compact_level(index_rel, 0);
 	}
```
[the remark this answers is not in the document]
Comment (tjgreen42) [reply src/access/build.c:537]: Fixed in b5586403 — bm25_force_merge now acquires the per-index LW_EXCLUSIVE before driving the merge and releases after. Concurrent backend scans hold LW_SHARED so they block for the duration of the force-merge, which is acceptable given force-merge is an administrative operation. Also added a bounds-check on hdr->level in tp_redo_apply_merge_linkage (separate review item from the same comment about pre-existing weak validation; the primary already gates emission, but redo must independently validate against corrupted / old / version-mismatched WAL records).

**Comment by tjgreen42** [comment]
Addressed in b5586403:

- 🔴 **Out-of-bounds level in MERGE_LINKAGE redo** — Added a `hdr->level >= TP_MAX_LEVELS - 1` bounds check at the top of `tp_redo_apply_merge_linkage` (PANIC on violation, ahead of any mutation). Independent of the primary's emit-time gate, since redo must validate against corrupted / old / version-mismatched WAL records on its own.
